server/src-lib/Hasura/GraphQL/Execute/Query.hs

```diff
@@ -135,56 +130,49 @@ mkCurPlanTx
   -> HTTP.Manager
   -> [HTTP.Header]
   -> UserInfo
-  -> [G.Directive G.Name]
-  -> FieldPlans
-  -> m (tx EncJSON, GeneratedSqlMap)
-mkCurPlanTx env manager reqHdrs userInfo directives fldPlans = do
+  -> (Q.Query -> Q.Query)
+  -> ExtractProfile
+  -> RootFieldPlan
+  -> (tx EncJSON, Maybe PreparedSql)
+mkCurPlanTx env manager reqHdrs userInfo instrument ep = \case
   -- generate the SQL and prepared vars or the bytestring
-  resolved <- forM fldPlans $ \(alias, fldPlan) -> do
-    fldResp <- case fldPlan of
-      RFPRaw resp                      -> return $ RRRaw resp
-      RFPPostgres (PGPlan q _ prepMap remoteJoins) -> do
-        let args = withUserVars (_uiSession userInfo) $ IntMap.elems prepMap
-        return $ RRSql $ PreparedSql q args remoteJoins
-      RFPActionQuery tx -> pure $ RRActionQuery tx
-    return (alias, fldResp)
-
-  (,) <$> mkLazyRespTx env manager reqHdrs userInfo directives resolved <*> pure (mkGeneratedSqlMap resolved)
+    RFPPostgres ps@(PreparedSql q prepMap remoteJoinsM) ->
+      let args = withUserVars (_uiSession userInfo) prepMap
+          -- WARNING: this quietly assumes the intmap keys are contiguous
+          prepArgs = fst <$> IntMap.elems args
+      in (, Just ps) $ case remoteJoinsM of
+           Nothing -> do
+             Tracing.trace "Postgres" . (runExtractProfile ep =<<) . liftTx $ asSingleRowJsonResp (instrument q) prepArgs
```

Comment:
This is another instance of an operator section involving `=<<`, which I missed on the first pass, but I personally find confusing. There’s a lot going on all on one line here. I recognize this kind of thing is subjective, but I happen to prefer a formulation like this:

```suggestion
             Tracing.trace "Postgres" $ runExtractProfile ep =<< liftTx do
               asSingleRowJsonResp (instrument q) prepArgs
```

This puts all the “wrapping bits” on a separate line from the actual interesting functionality being executed. Another approach I use sometimes when the “wrapping bits” get complicated is something like this:

```suggestion
             asSingleRowJsonResp (instrument q) prepArgs
               &   liftTx
               >>= runExtractProfile ep
               &   Tracing.trace "Postgres"
```

Which has sort of a “pipeline” flavor to it. This works out nicely because `&` and `>>=` are both `infixl 1`. But in this particular case, I’m not sure that it’s actually better; most of the parts of the “pipeline” are pretty boring.

Reply:
Done!